### PR TITLE
Implement viz_publish_service lambda auto republish via geoprocessing service

### PR DESCRIPTION
**Pull request by EdisonOrellana-NOAA**
This PR is essentially the same as #630 but with a few changes and loose ends to tie up. I believe mostly everything is captured in the list below but I will add additional information as a PR comment as it comes up.

## Background
At this point, all services have been moved off of the Viz EC2 instance, and it only performs one task: converting ArcGIS Pro Projects (which we save as MapX documents) into the ArcGIS service definition (SD) files that viz pipelines need to automatically publish services to ArcGIS server, including updating the data sources with the correct database host and authentication information.

The Viz EC2 instance does this by running the` \aws_loosa\deploy\update_data_stores_and_sd_files.py` script as part of the `Core\EC2\viz\scripts\viz_ec2_setup.ps1` deployment script, and this is the number one cause of issues during all deployments, as it takes quite a bit of work to dig into the logs and troubleshoot issues on the Viz EC2 machine. Because ESRI is greedy and won't fully allow their Python API to update these map document parameters (or allow use of a limited version of ArcPy in a container that can do it), we've had to get creative on another way to automatically update these map service definitions as part of our automatic pipelines - credit to Justin for initially coming up with the idea of using a geoprocessing service.

With Bob's help, I've successfully tested this geoprocessing service in TI, and while questions remain around how this will actually work operationally with the regular pipelines, it seems worth a shot. That said, because it is ESRI, there is a pretty convoluted workflow required to make this happen.

## Steps required to finish

- [x] Convert the update_data_stores_and_sd_files python script into a standalone script that can be published as a ArcGIS Geoprcessing Service (in this PR).
- [x] Test the geoprocessing service via REST - you can find the test [here](https://maps-testing.water.noaa.gov/gp/rest/services/Utilities/ServiceRepublish/GPServer).
- [ ] Allow the ESRISSMDeploy IAM Role to view the EGIS RDS user secret (this allows the geoprocessing server to fetch the RDS password securely from AWS without the password being passed around in text in the REST calls). The Terraform for this is included in #630 
- [x] @nickchadwick-noaa Add IaC for copying *.mapx files in the subdirectories of `Core/LAMBDA/viz_functions/viz_publish_service/services` to the expected location on S3. Currently in ti this is `s3://hydrovis-ti-deployment-us-east-1/viz/pro_projects/`. This value is currently hard coded in the lambda_function.py but should be set from an environment variable.
- [x] Add additional lambda enviornment variables in terraform: <html><body>

EGIS_DB_DATABASE | hydrovis
-- | --
EGIS_DB_HOST | rds-egis.hydrovis.internal
EGIS_DB_PASSWORD_SECRET_NAME | hv-vpp-ti-egis-pg-rds-secret
EGIS_DB_USERNAME | hydrovis
ENVIRONMENT | ti

</body>
</html>

- [ ] ~~Configure ArcGIS Server to use a custom python environment with the AWS Boto3 library installed. Bob did this in TI following the instructions of[ this ESRI doc](https://enterprise.arcgis.com/en/server/latest/publish-services/windows/deploying-custom-python-packages.htm). I believe this will need to be baked into the AMI, so this task should be assigned to whoever owns that process. I'm not sure which server of the ArcGIS stack this should be installed on, probably the same one as the print service?~~ After futilely attempting to make a custom python environment work, @shawncrawley suggested avoiding the problem by making a `subprocess` call to the `aws` command to avoid needing boto3. This worked well. [See this comment](https://github.com/NOAA-OWP/hydrovis/issues/446#issuecomment-2307254263)
- [ ]  Publish the geoprocessing service to each ArcGIS Server stack - This is unfortunately a manual process that must be done with the admin credentials. See [these instructions](https://github.com/NOAA-OWP/hydrovis/pull/630#issuecomment-1947128348) which are more or less still the same. Exception is regarding default values and which values are required.
- [x] Update the viz_publish_service lambda function to call this geoprocessing service first via API when it needs to republish a service. (changes included in this PR) This can be done now in TI, using the test geoprocessing service I created, to test this out with the operational TI pipelines.
- [ ] Delete the Viz EC2 entirely (it should be manually stopped after a deployment for the time being, as it doesn't serve any purpose after the initial deployment steps).

## Testing

1. Additional unit tests per service?

### CODE REVIEW

File: Core/LAMBDA/viz_functions/viz_publish_service/gp_service/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+{\rtf1}
```
Comment (shawncrawley) [inline Core/LAMBDA/viz_functions/viz_publish_service/gp_service/README.md:1]: What is the purpose of this?

Comment (EdisonOrellana-NOAA) [reply Core/LAMBDA/viz_functions/viz_publish_service/gp_service/README.md:1]: Remnant from creating a markdown file on Windows. (**r**ich **t**ext **f**ormat)
This readme is a placeholder meant to hold documentation for the GP service.

File: Core/LAMBDA/viz_functions/main.tf [hunk 2]
```diff
@@ -942,6 +950,11 @@ resource "aws_lambda_function" "viz_publish_service" {
       S3_BUCKET           = var.viz_authoritative_bucket
       SD_S3_PATH          = "viz_sd_files/"
       SERVICE_TAG         = local.service_suffix
+      EGIS_DB_HOST        = var.egis_db_host
+      EGIS_DB_DATABASE    = var.egis_db_name
+      EGIS_DB_USERNAME    = jsondecode(var.egis_db_user_secret_string)["username"]
+      EGIS_DB_PASSWORD    = jsondecode(var.egis_db_user_secret_string)["password"]
```
Comment (EdisonOrellana-NOAA) [inline Core/LAMBDA/viz_functions/main.tf:956]: Update lambda to read environment variable "EGIS_DB_PASSWORD" instead of "EGIS_DB_PASSWORD_SECRET_NAME"

**Comment by EdisonOrellana-NOAA** [comment]
Thanks for the help getting this on UAT!
Further steps:
- [x] Update the URL to call the service in UAT. This gets updated in the Lambda
- [x] Fix the terraform to configure the proper location for the mapx files

**Review by nickchadwick-noaa — DISMISSED** [review]

**Review by nickchadwick-noaa — APPROVED** [review]

**Comment by EdisonOrellana-NOAA** [comment]
- [x] Change the service name on maps-testing to match what is named in UAT
- [ ] Delete the parts that are currently creating SD files since this service deprecates that functionality
- [x] Update read me to match staging
- [ ] Follow up with Shawn about examining success on UAT
- [ ] I left the service with the old name alongside it. So nothing is broken currently. After successful deployment to TI, it should start using the new service and then we can [delete the old service.](https://maps-testing.water.noaa.gov/gp/admin/services/Utilities/ServiceRepublish.GPServer/delete)

**Comment by EdisonOrellana-NOAA** [comment]
# UAT test requested for geoprocessing service
The geoprocessing service is meant to convert a mapx file to a service definition (.sd) file and place the file in the necessary location on S3. [The viz_publish_service lambda publishes the service using that .sd file.](https://github.com/NOAA-OWP/hydrovis/blob/ti/Core/LAMBDA/viz_functions/viz_publish_service/lambda_function.py#L101)
## Test procedure
1. Rename/move a service definition file. Example, _ana_high_flow_magnitude_ak_noaa.sd_
`s3://hydrovis-uat-deployment-us-east-1/viz_sd_files/ana_high_flow_magnitude_ak_noaa.sd`
2. Rename/move the corresponding publish flag file. Example, _ana_high_flow_magnitude_ak_noaa_
`s3://hydrovis-uat-fim-us-east-1/published_flags/server/nwm/ana_high_flow_magnitude_ak_noaa/ana_high_flow_magnitude_ak_noaa`
3. Either trigger a pipeline run in UAT that includes _ana_high_flow_magnitude_ak_noaa_ (preferred) OR trigger the GP service manually at the [REST URL](https://maps-staging.water.noaa.gov/gp/rest/services/Utilities/MapxToSD/GPServer/MapxToSD/execute)
4. Observe Cloudwatch logs for the Lambda: `hv-vpp-uat-viz-publish-service` and look for the log entry: `---> ana_high_flow_magnitude_ak_noaa does not currently exist. Attempting to publish via GP service.`
5. Verify successful recreation of the service definition (.sd) file at `s3://hydrovis-uat-deployment-us-east-1/viz_sd_files/ana_high_flow_magnitude_ak_noaa.sd`
6. Verify successful recreation of publish flag at `s3://hydrovis-uat-fim-us-east-1/published_flags/server/nwm/ana_high_flow_magnitude_ak_noaa/ana_high_flow_magnitude_ak_noaa`
7. Verify the tested service is functional in UAT
### Success
.sd and publish flag are successfully recreated and service is successfully republished and working
### Failure
.sd and publish flag are **NOT** successfully recreated and service is successfully republished and working
#### Recovery/rollback
Manually restore the .sd file and publish flag in their respective locations in S3

**Pull request merged**